## Metrics - Overview

Useful to evaluate how precise and accurate a ml model is.<br> 
Two types of evaluation metrics are there:<br>

i. Regression metrics<br>
ii. Classification metrics

# Regression metrics

a. Mean squared error(mse) : Average squared difference between predicted and actual values.

MSE = (1/n) * Σ (predicted value - actual value)² <br>

n is the number of data points.<br>
Σ represents the summation of all squared differences.<br>

In [1]:
# importing necessary libraries 
from sklearn.linear_model import LinearRegression 
import seaborn as sb 
from sklearn.model_selection import train_test_split 
from sklearn.metrics import mean_squared_error
# loading the dataset
df = sb.load_dataset('tips') 
# selecting features and target variable 
X = df[['total_bill', 'size']]
y = df['tip'] 

In [2]:
# splitting the dataset into training and testing sets 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) 
# creating a linear regression model
model = LinearRegression() 
# fitting the model to the training data
model.fit(X_train, y_train) 
# making predictions on the test data 
y_pred = model.predict(X_test) 
# displaying the made predictions 
y_pred[:5]

array([2.89943693, 1.89520002, 3.85933994, 3.9811207 , 2.27962736])

In [3]:
# calculating the mean squared error 
mse = mean_squared_error(y_test, y_pred) 
mse

0.6485996190543517

<b>Interpretation:</b> The average error or deviation of predicted value from the original values is about 0.65 according to mse(mean squared error) indicating good model performance but still room for improvement.

In [4]:
# calculating root mean squared error through custom function 
import numpy as np
def root_mean_squared_error(y_true, y_pred):
    return np.sqrt(mean_squared_error(y_true, y_pred)) 
rmse = root_mean_squared_error(y_test, y_pred) 
rmse

np.float64(0.8053568271607013)

b. Mean absolute error(mae) : calculates the average of the absolute differences between predicted and actual values along with providing a straightforward way to quantify the errors.<br>

MAE = (1/n) * Σ |predicted value - actual value|<br>

n is the number of data points.<br>
Σ represents the summation of all squared differences.<br>

In [5]:
# calculating the mean absolute error 
from sklearn.metrics import mean_absolute_error 
mae = mean_absolute_error(y_test, y_pred) 
mae

0.6639235737596481

<b>Interpretation:</b> The average error or deviation of predicted value from the original values is about 0.65 according to mae(mean absolute error) indicating good model performance but still room for improvement.

c. Mean absolute percentage error(mape):  expresses the average percentage difference between predicted and actual values providing a relative measure of error which is very easy to understand.<br>

MAPE = (1/n) * Σ |(actual value - predicted value) / actual value| * 100<br>

n is the number of data points.<br>
Σ represents the summation of all squared differences.<br>

In [6]:
# calculating the mean absolute percentage error from custom made function 
def mean_absolute_percentage_error(y_true, y_pred): 
    y_true, y_pred = np.array(y_true), np.array(y_pred) 
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100
mape = mean_absolute_percentage_error(y_test, y_pred) 
mape

np.float64(27.978917485910255)

<b>Interpretation:</b> According to mean absolute percentage error(mape) about 28% of the predictions done by the model is errorful.

d. Mean squared logarithmic error(msle) : Improved version of mean sqaured error(mse) but with logarithmic functions when the data has wide(varied) range of values.<br>

MSLE = (1/n) * Σ [log(predicted_value - actual value)] <br>

n is the number of data points.<br>
Σ represents the summation of all squared differences.<br>

In [7]:
# calculating the mean sqaured logarithmic error from custom made function 
def mean_squared_logarithmic_error(y_true, y_pred): 
    y_true, y_pred = np.array(y_true), np.array(y_pred) 
    return np.mean(np.square(np.log1p(y_pred) - np.log1p(y_true)))
msle = mean_squared_logarithmic_error(y_test, y_pred) 
msle

np.float64(0.04365490227864684)

<b>Interpretation:</b> mean sqaured logarithmic error(msle) is 0.04 determining the ability of the model to handle most varied values in the dataset. 

e. r2_score : Determines the proportion of the variance in the dependent variable that is predictable from the independent variables. A good r2_score is often considered to be in range of 70 to 99.<br>

In [8]:
# calculating r2_score 
from sklearn.metrics import r2_score 
r2 = r2_score(y_test, y_pred) 
r2

0.4811084097989491

<b>Interpretation:</b> The r2_score is only 0.48 which is non satisfactory indicating the model's inability to predict the dependent variable('tip') based on two independent variables('total_bill' & 'size'). Only 48% of the total variance in the dependent variable is predicted through the independent variables.

# Classification metrics

a. accuracy_score : evaluates the exact state of matching between the predicted and actual labels.<br>

Binary classification : accuracy_score = (True Positives + True Negatives) / (True Positives + True Negatives + False Positives + False Negatives) <br>

True Positives (TP): Correctly predicted positive.<br>
True Negatives (TN): Correctly predicted negative.<br>
False Positives (FP): Incorrectly predicted positive (Type I error).<br>
False Negatives (FN): Incorrectly predicted negative (Type II error).<br>

Multilabel classification : accuracy_score = (Number of Correct Predictions) / (Total Number of Predictions)


In [9]:
# load the dataset 
df = sb.load_dataset('titanic') 
# selecting features and target variable for binary classification
df = df.dropna(subset=['age', 'fare', 'survived'])  # dropping rows with NaN values in selected columns
X = df[['age', 'fare']] 
y = df['survived'] 
# splitting the dataset into training and testing sets 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# creating a logistic  regression model
from sklearn.linear_model import LogisticRegression 
model = LogisticRegression() 
# fitting the model to the training data 
model.fit(X_train, y_train)
# making predictions on the test data 
y_pred = model.predict(X_test) 
# displaying the made predictions
y_pred[:5]

array([0, 1, 0, 1, 0])

In [ ]:
from sklearn.metrics import accuracy_score,recall_score, f1_score  # precision_score will be implemented with custom function
# calculating accuracy 
accuracy = accuracy_score(y_test, y_pred) 
accuracy

0.6293706293706294

<b>Interpretation:</b> According to the accuracy_score of about 0.63 (~0.62) , about 63 times out of 100 the label is predicted correctly

b. precision_score : measures the accuracy of positive predictions.

precision_score = True Positives / (True Positives + False Positives)

In [ ]:
# implementing precision score with custom function 
def precision_score(y_true, y_pred):
    true_positive = np.sum((y_true == 1) & (y_pred == 1))
    false_positive = np.sum((y_true == 0) & (y_pred == 1))
    return true_positive / (true_positive + false_positive) if (true_positive + false_positive) > 0 else 0
precision = precision_score(y_test, y_pred)
precision

np.float64(0.5483870967741935)

<b>Interpretation</b> According to the precision_score of 54.83, about 55 percent of the test data is predicted successfully.

c. recall_score : a metric to evaluate how well a model identifies positive instances.<br>

recall_score = TP / (TP + FN)<br>

TP : True Positives<br>
FN : False Negatives<br>

In [12]:
# calculating recall score 
recall = recall_score(y_test, y_pred)
recall

0.30357142857142855

<b>Interpretation:</b> recall_score of 0.30 shows that only 30% of the positives are identified by the model.